project/src/test/java/com/google/sps/servlets/QueryServletTest.java

```diff
@@ -67,36 +72,122 @@ public void getRequest_fetchedMoreThanMaxNumPlaces_respondMaxNumPlaces() throws
+    Place lowRating = createValidPlaceBuilder().setRating(3).setName("lowRatingPlace").build();
+    Place noWebsite = createValidPlaceBuilder().setWebsiteUrl("").setName("noWebsitePlace").build();
+    // The following place purposely has the same name as the valid place, so that we make sure that
+    // we end up with only one of them.
+    Place validPlace2 = createValidPlaceBuilder().setName("validPlace").build();
+    ImmutableList<Place> places =
+        ImmutableList.of(validPlace, lowRating, noWebsite, validPlace2);
+    when(FETCHER.fetch(any(UserPreferences.class))).thenReturn(places);
+
+    servlet.doGet(REQUEST, RESPONSE);
+
+    assertEquals(getPlacesAmountInResponse(), 1);
+    assertEquals(
+        new Gson().fromJson(responseStringWriter.getBuffer().toString(), JsonArray.class)
+            .get(0) // The first element in the array (first place)
+            .getAsJsonObject()
+            .get("name")
+            .getAsString(),
+        "validPlace");
+  }
+
+  @Test
+  public void getRequest_fetcherException_forwardException() throws Exception {
+    when(FETCHER.fetch(any(UserPreferences.class))).thenThrow(FetcherException.class);
+
+    servlet.doGet(REQUEST, RESPONSE);
+
+    verify(RESPONSE).sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+        "Fetching from Google Places API encountered a problem");
+  }
+
+  @Test
+  public void getRequest_userPreferencesBuildException_forwardException() throws Exception {
+    // The rating parameter cannot be negative, so this should cause the UserPreferences builder to
+    // throw an IllegalArgumentException
+    when(REQUEST.getParameter("rating")).thenReturn("-5");
+
+    servlet.doGet(REQUEST, RESPONSE);
+
+    verify(RESPONSE).sendError(HttpServletResponse.SC_BAD_REQUEST,
+        "Parsing the user preferences encountered a problem");
+  }
+
+  @Test
+  // This test checks that the user preferences are processed correctly based on the given
+  // parameters. The crucial point of the test is that the strings that are given as parameters
+  // match the expect user preferences.
+  public void getRequest_userPreferencesForwardedToFetcher() throws Exception {
+    when(REQUEST.getParameter("rating")).thenReturn("4");
+    when(REQUEST.getParameter("price")).thenReturn("3");
+    when(REQUEST.getParameter("open")).thenReturn("1");
+    when(REQUEST.getParameter("location")).thenReturn("30.30000000,35.35000000");
+    when(REQUEST.getParameter("cuisines")).thenReturn("sushi,hamburger");
+    UserPreferences expectedUserPrefs = UserPreferences.builder()
+        .setMinRating(4)
+        .setMaxPriceLevel(3)
+        .setOpenNow(true)
+        .setLocation(new LatLng(30.30000000, 35.35000000))
+        .setCuisines(ImmutableList.of("sushi", "hamburger"))
+        .build();
+
+    servlet.doGet(REQUEST, RESPONSE);
+
+    // TODO(M1): this test fails because float is unstable for that kind of a test.
```

Comment:
I prefer that you simply make a simple float like "0". Doesn't this work? 
And then the TODO will be to change the float.

Reply:
Great idea, this does work. done.